### PR TITLE
[Feature] Add a section list and changePassword Modal to Setting screen

### PR DESCRIPTION
**Pull request by JongtaekChoi**
## Description

I added the section list and section item interface to the Setting screen. And I implement the changing password modal. 

## Related Issues

#29 

## Tests

I added the following tests:

- render Setting screen 
- press login state  section item 
- render ChangePwScreen
- press change password button
- text change of password text inputs 

## Checklist

Before you create this PR confirms that it meets all requirements listed below by checking the relevant checkboxes (`[x]`). This will ensure a smooth and quick review process.

- [x] I read the [Contributor Guide](https://github.com/dooboolab/hackatalk-mobile/blob/master/CONTRIBUTING.md) and followed the process outlined there for submitting PRs.
- [x] I signed the [CLA].
- [x] Run `yarn test` or `yarn test -u` if you need to update snapshot.
- [x] Run `yarn lint`
- [x] I am willing to follow-up on review comments in a timely manner.

### CODE REVIEW

File: src/components/screen/Setting/ChangePwView.tsx [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+}
+
+function ChangePwView(props: Props): React.ReactElement {
+  const { theme, close, checkCurrentPw } = props;
+  const [currentPwVerified, setCurrentPwVerified] = useState(false);
+  const [currentPw, setCurrentPw] = useState('');
+  const [newPw, setNewPw] = useState('');
+  const [newPwCheck, setNewPwCheck] = useState('');
+  const checkCurrent = (): void => {
+    checkCurrentPw().then((verified) => {
```
Comment (Sandwichj) [inline src/components/screen/Setting/ChangePwView.tsx]: The naming `check` means A, B or more.
And this function has two roles `Updating state` and `Validating password`.
So, if the result value is `boolean`, the roles should be separated for readability and intuition.

For example,
```js
isValid~()  // Validate password
updated~Verified(isValid~()) // Update state
```

It makes code clear. (ref. [Single Responsibility Principle](https://en.wikipedia.org/wiki/Single_responsibility_principle))

**Comment by JongtaekChoi** [comment]
> Try to fill out the description on your `PR`

I did it!

File: src/components/screen/ProfileUpdate.tsx [hunk 2]
```diff
@@ -79,6 +79,12 @@ function Screen(props: Props): React.ReactElement {
     setIsUpdating(true);
   };
 
+  const onSetting = (): void => {
+    if (props.navigation) {
```
Comment (hyochan) [inline src/components/screen/ProfileUpdate.tsx]: It'd be better to use `const navigation = useNavigation();`.

File: src/components/screen/Setting/ChangePwView.tsx [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+import { InnerContainer, ModalCloseButton, ModalHeader, ModalTitle, StyledTextInput } from './styles';
+import React, { useState } from 'react';
+
+import Button from '../../shared/Button';
+import { Ionicons } from '@expo/vector-icons';
+import { getString } from '../../../../STRINGS';
+
+export interface Props {
+    theme: DefaultTheme;
+    close(): void;
```
Comment (hyochan) [inline src/components/screen/Setting/ChangePwView.tsx]: I don't think this is a good idea if this is a `screen` component. It'd be better to pass the param to `react-navigation` and get it from the `route` param.

File: src/components/screen/Setting/ChangePwView.tsx [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+    validateCurrentPw(text: string): Promise<boolean>;
```
Comment (hyochan) [inline src/components/screen/Setting/ChangePwView.tsx]: same here.

File: src/components/screen/Setting/ChangePwView.tsx [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+} from 'styled-components/native';
+import { InnerContainer, ModalCloseButton, ModalHeader, ModalTitle, StyledTextInput } from './styles';
+import React, { useState } from 'react';
+
+import Button from '../../shared/Button';
+import { Ionicons } from '@expo/vector-icons';
+import { getString } from '../../../../STRINGS';
+
+export interface Props {
+    theme: DefaultTheme;
```
Comment (hyochan) [inline src/components/screen/Setting/ChangePwView.tsx]: This is currently deprecated. We are using `const theme = useThemeContext()`.

File: src/components/screen/Setting/ChangePwView.tsx (hunk 4 above)
Comment (hyochan) [reply src/components/screen/Setting/ChangePwView.tsx]: What screen prop has is simply below.
```
interface Props extends ThemeProps<DefaultTheme> {
  navigation: DefaultNavigationProps;
}
```

I currently don't think there could be more than this.

File: src/components/screen/Setting/index.tsx [hunk 6]
```diff
@@ -0,0 +1,87 @@
...
+    <ItemContainer onPress={option.onPress} testID={option.testID}>
+      {option.icon ? <ItemIcon source={option.icon} /> : null}
+      <ItemLabel>{option.label}</ItemLabel>
+      <FontAwesome name="angle-right" size={32} color={theme.fontColor} />
+    </ItemContainer>
+  );
+}
+
+interface Props {
+  navigation: DefaultNavigationProps<'Setting'>;
```
Comment (hyochan) [inline src/components/screen/Setting/index.tsx]: Good one here. You are using it correctly.

File: src/components/screen/Setting/index.tsx [hunk 7]
```diff
@@ -0,0 +1,87 @@
...
+  theme: DefaultTheme;
```
Comment (hyochan) [inline src/components/screen/Setting/index.tsx]: As said above, this can be removed.

File: src/components/screen/Setting/index.tsx [hunk 8]
```diff
@@ -0,0 +1,87 @@
...
+}
+
+function SettingScreen(props: Props): React.ReactElement {
+  const { theme } = props;
```
Comment (hyochan) [inline src/components/screen/Setting/index.tsx]: Replace this with `const { theme } = useThemeContext();`

File: src/components/screen/Setting/index.tsx [hunk 9]
```diff
@@ -0,0 +1,87 @@
...
+  const modal: React.MutableRefObject<Modal|null> = useRef(null);
+
+  const settings: SectionListData<SettingsOption>[] = [
+    {
+      title: '계정정보',
```
Comment (hyochan) [inline src/components/screen/Setting/index.tsx]: Try to localize this with `getString` since this looks constant.

File: src/components/screen/Setting/index.tsx [hunk 10]
```diff
@@ -0,0 +1,87 @@
...
+      data: [{
+        icon: IC_FACEBOOK,
+        label: 'facebook으로 로그인 됨',
```
Comment (hyochan) [inline src/components/screen/Setting/index.tsx]: Same here! Localize~

File: src/components/screen/ProfileUpdate.tsx (hunk 2 above)
Comment (JongtaekChoi) [reply src/components/screen/ProfileUpdate.tsx]: Actually it's now unnecessary. I will remove it. Thanks.

File: src/components/screen/Setting/index.tsx (hunk 6 above)
Comment (JongtaekChoi) [reply src/components/screen/Setting/index.tsx]: Thanks. But, is it needed? In this https://github.com/dooboolab/hackatalk-mobile/pull/58#discussion_r347398884 you said that it'd be better to use `const navigation = useNavigation();`

File: src/components/screen/Setting/ChangePwView.tsx (hunk 5 above)
Comment (JongtaekChoi) [reply src/components/screen/Setting/ChangePwView.tsx]: I'd remove it.

File: src/components/screen/Setting/index.tsx (hunk 7 above)
Comment (JongtaekChoi) [reply src/components/screen/Setting/index.tsx]: I'd remove it. Thanks.

File: src/components/screen/Setting/index.tsx (hunk 8 above)
Comment (JongtaekChoi) [reply src/components/screen/Setting/index.tsx]: I will change it. Thx.

File: src/components/screen/Setting/index.tsx (hunk 6 above)
Comment (hyochan) [reply src/components/screen/Setting/index.tsx]: This is needed if you want to strictly check the state param passed down to the `Setting` screen.

Oh yes, and sorry for confusing you. I wanted to tell you that `useNavigation` would be efficient in such cases if you need the navigation props in `shared` components. Then you won't have to pass down the `navigation` props every time. In this case, `useNavigation` is not a good idea.

**Comment by JongtaekChoi** [comment]
@hyochan I added some changes.
1. ChangePwView was moved to the shared folder and renamed to ChangePwModal
2. Do I need to remove the navigation options of the 'ProfileUPdate' stack screen?

File: src/components/screen/Setting/index.tsx [hunk 11]
```diff
@@ -0,0 +1,100 @@
...
+          </HeaderContainer>
+        )}
+      />
+      <ChangePwModal
+        ref={modal} />
+    </Container>
+  );
+}
+
+export default withTheme(SettingScreen);
```
Comment (hyochan) [inline src/components/screen/Setting/index.tsx]: You don't need `withTheme` with hook.

File: src/components/screen/Setting/styles.ts [hunk 12]
```diff
@@ -0,0 +1,72 @@
...
+width: 100%;
+height: 52px;
+align-items: center;
+/* border-width: 1px; */
+`;
+
+export const ItemIcon = styled.Image`
+width: 32px;
+height: 32px;
+resize-mode: contain;
```
Comment (hyochan) [inline src/components/screen/Setting/styles.ts]: This isn't a style!

File: src/components/shared/ChangePwModal.tsx [hunk 13]
```diff
@@ -0,0 +1,133 @@
...
+      keyboardTopOffset={0}
+      coverScreen
+      backButtonClose
+      style={{ backgroundColor: theme.background }}>
+      <ChangePw close={close} />
+    </Modal>
+  );
+});
+
+export default withTheme(ModalWrapper);
```
Comment (hyochan) [inline src/components/shared/ChangePwModal.tsx]: No need to wrap with `withTheme` in hook!

**Comment by JongtaekChoi** [comment]
@hyochan Thanks for your work. I have checked the problem of 'blinking'. It occurred when closing the ChangePwModal. I'll try to fix this.

**Comment by hyochan** [comment]
@JongtaekChoi How is this going?

**Comment by JongtaekChoi** [comment]
I think it caused by SafeAreaView.
When I changed SafeAreaView to normal View it doesn't occur.

image
[refer to Reat Navigation Docs](https://reactnavigation.org/docs/en/next/handling-safe-area.html)

Fortunately, [expo support `react-native-safe-area-context` library.](https://docs.expo.io/versions/v35.0.0/sdk/safe-area-context/)

I will try this method within today and add commit if it works.

> @JongtaekChoi How is this going?

**Comment by JongtaekChoi** [comment]
> I think it caused by SafeAreaView.
> When I changed SafeAreaView to normal View it doesn't occur.
> 
> image
> [refer to Reat Navigation Docs](https://reactnavigation.org/docs/en/next/handling-safe-area.html)
> 
> Fortunately, [expo support `react-native-safe-area-context` library.](https://docs.expo.io/versions/v35.0.0/sdk/safe-area-context/)
> 
> I will try this method within today and add commit if it works.
> 
> > @JongtaekChoi How is this going?

It doesn't work.

**Comment by hyochan** [comment]
@JongtaekChoi Sad news.. Then what if changing your modal into screen component?

**Comment by JongtaekChoi** [comment]
I'll try a little more.

**Comment by JongtaekChoi** [comment]
@hyochan  I've made a change to the headers of UpdateProfile and Setting screen. 
I found that the Chat and SearchUser screens have the same issues and I fixed them. 
It may be a temporal solution and some similar issues may exist. 
But I think this pr is enough to long and these issues could be fixed at another pr or commit.

File: src/components/screen/ChangePw/index.tsx [hunk 14]
```diff
@@ -0,0 +1,131 @@
...
+        <CloseButton testID="closeBtn" onPress={close}>
+          <View
+            style={{
+              padding: 8,
+            }}
+          >
+            <Ionicons name="md-close" size={24} color={theme.fontColor} />
+          </View>
+        </CloseButton>
+      </Header>
```
Comment (hyochan) [inline src/components/screen/ChangePw/index.tsx]: I think removing this part will prevent such side effects on the header. Try using header which is already implemented in `react-navigation` that will handle such cases.

**Comment by JongtaekChoi** [comment]
@hyochan Thanks for your review. 
I made it by adding options to the ChangePw screen stack.
And I set `headerMode` of MainStackNavigator to 'screen' because of this [document](https://reactnavigation.org/docs/en/next/stack-navigator.html#header). 
image
It makes the header attached to the screen. 
so the bug you mentioned at [here](https://github.com/dooboolab/hackatalk-mobile/pull/58#pullrequestreview-318349931) also be resolved by this `screen` option. 
Dec-19-2019 12-25-46

So I request you to review my codes. 
Thanks again.

File: src/components/navigation/AuthStackNavigator.tsx [hunk 15]
```diff
@@ -11,26 +11,28 @@ const Stack = createStackNavigator();
...
-      initialRouteName="Login"
-      screenOptions={{
-        headerStyle: {
-          backgroundColor: theme.background,
-          borderBottomWidth: 0,
-          elevation: 0,
-        },
-        headerTintColor: theme.fontColor,
-      }}
-    >
```
Comment (hyochan) [inline src/components/navigation/AuthStackNavigator.tsx]: @JongtaekChoi Why did you remove this? It will remove the dark mode behavior in the header as shown below?

image

Comment (JongtaekChoi) [reply src/components/navigation/AuthStackNavigator.tsx]: @hyochan Oh it's my mistake. 
When I was finding the cause of error about the header option, I tested if the secreenOptions be removed then the error occurred. And I forgot to revert this.
